sdks/go/pkg/beam/runners/prism/internal/jobservices/management.go

```diff
@@ -243,8 +243,8 @@ func (s *Server) Run(ctx context.Context, req *jobpb.RunJobRequest) (*jobpb.RunJ
 // Otherwise, returns nil if Job does not exist or the Job's existing state as part of the CancelJobResponse.
 func (s *Server) Cancel(_ context.Context, req *jobpb.CancelJobRequest) (*jobpb.CancelJobResponse, error) {
 	s.mu.Lock()
+	defer s.mu.Unlock()
```

Comment:
deferring unlock isn't a straightforward universal good.

So this locks the *whole server's shared state* until the Cancel is complete, while it's doing the per job changes. You would at best want to lock the job's state, not serialize everything the server is handling.

The various job state fields should be locked through a job specific lock, which I suppose might be job.streamCond.L. State in particular is an atomic to avoid needing a full job lock for every read. With cancelling that might need to change however.

The Prepare method definitely has some overzealous locking (it's doing a lot of work, but not a lot of mixing with the server side state.

Reply:
Now it makes sense to me that state is an atomic.Value and thus I didn't need to put a lock on the whole service. I only lock/unlock when acquiring the Job just like all the other methods.